utils/vlog/vlog_to_pbtype.py

```diff
@@ -0,0 +1,210 @@
+            pb_xml_attrs["blif_model"] = ".latch"
+            pb_xml_attrs["class"] = "flipflop"
+        else:
+            assert False
+    elif mod_type == "blackbox":
+        pb_xml_attrs["blif_model"] = ".subckt " + mod.get_name()
+        if mod_cls is not None:
+            if mod_cls == "mem":
+                pb_xml_attrs["class"] = "memory"
+            else:
+                assert False
+    else:
+        assert False
+
+    #TODO: might not always be the case? should be use Verilog `generate`s to detect this?
+    pb_xml_attrs["num_pb"] = "1"
+
+    if xml_parent is None:
+        pb_type_xml = ET.Element("pb_type", pb_xml_attrs)
+    else:
+        pb_type_xml = ET.SubElement(xml_parent, "pb_type", pb_xml_attrs)
+
+    # List of entries in format ((from_cell, from_pin), (to_cell, to_pin))
+    interconn = []
+    # List of interconnect to ignore, because it was generated inside a <mode>
+    ignore_interconn = []
+
+    # Process IOs
+    clocks = yosys.run.list_clocks(args.infiles, mod.get_name())
+    for name, width, iodir in mod.get_ports():
+        ioattrs = {"name": name, "num_pins": str(width), "equivalent": "false"}
+        if name in clocks:
+            ET.SubElement(pb_type_xml, "clock", ioattrs)
+        elif iodir == "input":
+            ET.SubElement(pb_type_xml, "input", ioattrs)
+        elif iodir == "output":
+            ET.SubElement(pb_type_xml, "output", ioattrs)
+        else:
+            assert False
+    # Process cells. First build the list of cnames.
+    for cname, i_of in mod.get_cells(False):
+        cname_to_pb_name[cname] = mod.get_cell_attr(cname, "PB_NAME", i_of)
+
+    for cname, i_of in mod.get_cells(False):
+        pb_name = mod.get_cell_attr(cname, "PB_NAME", i_of)
+        module_path = os.path.dirname(yj.get_module_file(i_of))
+        pb_type_path = "%s/pb_type.xml" % module_path
+        pb_mode = mod.get_cell_attr(cname, "MODE", "")
+        # If MODE is used, need to wrap cell pb_type and interconnect in "mode"
+        is_modal = (pb_mode != "")
+        if is_modal:
+            mode_xml = ET.SubElement(pb_type_xml, "mode", {'name': pb_mode})
+            ET.SubElement(mode_xml, "{http://www.w3.org/2001/XInclude}include", {'href': pb_type_path})
+            mode_ic = ET.SubElement(mode_xml, "interconn")
+        else:
+            ET.SubElement(pb_type_xml, "{http://www.w3.org/2001/XInclude}include", {'href': pb_type_path})
+        # In order to avoid overspecifying interconnect, there are two directions we currently
+        # consider. All interconnect going INTO a cell, and interconnect going out of a cell
+        # into a top level output - or all outputs if "mode" is used.
+        inp_cons = mod.get_cell_conns(cname, "input")
+        for pin, net in inp_cons:
+            drvs = mod.get_net_drivers(net)
+            if len(drvs) == 0:
+                print("WARNING: pin %s.%s has no driver, interconnect will be missing" % (pb_name, pin))
```

Comment:
This is pretty normal with multiple modes. Maybe this warning only shows up when not generating a mode thingy?